opal/runtime/opal_progress.c

```diff
@@ -254,6 +341,48 @@ int opal_progress(void)
     return events;
 }
 
+#if OPAL_ENABLE_PROGRESS_THREADS == 1
+static void *opal_progress_async_thread_engine(opal_object_t *obj)
+{
+    opal_thread_t *current_thread = (opal_thread_t *) obj;
+    thread_args_t *p_thread_arg = (thread_args_t *) current_thread->t_arg;
+
+    while (p_thread_arg->running) {
+        const int64_t new_events = _opal_progress();
+        if (new_events > 0) {
+            opal_atomic_add_fetch_64(&p_thread_arg->nb_events_reported, new_events);
+        }
+    }
+
+    return OPAL_THREAD_CANCELLED;
+}
+#endif
+
+int opal_progress(void)
```

Comment:
I don't think it makes sense to track and return the number of event managed by the progress thread. The reason is that it will mostly be inaccurate if there are multiple threads calling into `opal_progress` to check for progress, and that would lead them to make incorrect assumptions (because they will not see events that were `opal_atomic_swap_64` by another thread).

But this is not really an issue, it only affects a single location `ompi/request/req_test.c`, and the impact will be minimal. I propose you remote all atomic operations around `thread_arg.nb_events_reported`, and make `opal_progress` return 0 in all cases. Then, in `ompi/request/req_test.c` you always `goto recheck_request_status` after the call to `opal_progress`.

Reply:
I added atomic because the existing semantic of `int opal_progress()` is to return the number of processed events. And yes, the only places where the returned value is used seems to be `req_test.c`. 

If we proceed to what you recommend, I think it is better to change the function to `void opal_progress()` to be consistent to the semantic change ? What do you think ?

Reply:
> The reason is that it will mostly be inaccurate if there are multiple threads calling into opal_progress to check for progress, and that would lead them to make incorrect assumptions (because they will not see events that were opal_atomic_swap_64 by another thread).

Don't we have the same problem without the progress thread: two threads may call into `opal_progress` and one may progress some operation while the other doesn't. The second thread won't see that there actually was progress on operations.